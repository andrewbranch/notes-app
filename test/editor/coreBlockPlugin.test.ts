import { loadApp, typeText, getState, pressKey } from './transport';

const assertBlockType = async (blockType: string) => expect((await getState()).content.getFirstBlock().getType()).toBe(blockType);

beforeEach(async () => {
  await loadApp();
});

describe('coreBlockPlugin', () => {
  describe('headings', () => {
    test('doesn’t create a heading when typing hashes before space is pressed', async () => {
      await typeText('#');
      await assertBlockType('unstyled');
    });

    test('creates a heading after typing some hashes and a space', async () => {
      await typeText('# ');
      await assertBlockType('header-one');
      expect(await getState()).toMatchSnapshot();
    });

    test('creates a heading after typing a hash if there’s a trailing space', async () => {
      await typeText(' ');
      await pressKey('ArrowLeft');
      await typeText('#');
      await assertBlockType('header-one');
    });

    test('changes heading type when modifying hashes', async () => {
      await typeText('# ');
      await pressKey('ArrowLeft');
      await typeText('#');
      await assertBlockType('header-two');
      await pressKey('Backspace');
      await assertBlockType('header-one');
    });

    test('doesn’t create a heading if the hashes aren’t at the beginning of the block', async () => {
      await typeText(' # ');
      await assertBlockType('unstyled');
    });

    test('resets block type if the heading sequence becomes invalid', async () => {
      await typeText('## ');
      await assertBlockType('header-two');
      await pressKey('ArrowLeft');
      await typeText('x');
      await assertBlockType('unstyled');
      await pressKey('Backspace');
      await assertBlockType('header-two');
      await pressKey('ArrowLeft');
      await typeText('x');
      await assertBlockType('unstyled');
    });
  });
});
