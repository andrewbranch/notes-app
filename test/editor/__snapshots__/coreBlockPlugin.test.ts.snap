// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coreBlockPlugin decorator sequence styles removes other styles when styling decorator characters 1`] = `
Object {
  "content": Immutable.Record {
    "entityMap": Object {},
    "blockMap": Immutable.OrderedMap {
      "0": Immutable.Record {
        "key": "0",
        "type": "header-two",
        "text": "## ",
        "characterList": Immutable.List [
          Immutable.Record {
            "style": Immutable.OrderedSet [
              "core.block.decorator",
            ],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [
              "core.block.decorator",
            ],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
        ],
        "depth": 0,
        "data": Immutable.Map {},
      },
    },
    "selectionBefore": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
    "selectionAfter": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
  },
  "selection": Immutable.Record {
    "anchorKey": "0",
    "anchorOffset": 3,
    "focusKey": "0",
    "focusOffset": 3,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`coreBlockPlugin decorator sequence styles styles decorator characters when a block is created 1`] = `
Object {
  "content": Immutable.Record {
    "entityMap": Object {},
    "blockMap": Immutable.OrderedMap {
      "0": Immutable.Record {
        "key": "0",
        "type": "header-two",
        "text": "## ",
        "characterList": Immutable.List [
          Immutable.Record {
            "style": Immutable.OrderedSet [
              "core.block.decorator",
            ],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [
              "core.block.decorator",
            ],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
        ],
        "depth": 0,
        "data": Immutable.Map {},
      },
    },
    "selectionBefore": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
    "selectionAfter": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
  },
  "selection": Immutable.Record {
    "anchorKey": "0",
    "anchorOffset": 3,
    "focusKey": "0",
    "focusOffset": 3,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`coreBlockPlugin decorator sequence styles styles decorator characters when a block is expanded 1`] = `
Object {
  "content": Immutable.Record {
    "entityMap": Object {},
    "blockMap": Immutable.OrderedMap {
      "0": Immutable.Record {
        "key": "0",
        "type": "header-two",
        "text": "## Hello",
        "characterList": Immutable.List [
          Immutable.Record {
            "style": Immutable.OrderedSet [
              "core.block.decorator",
            ],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [
              "core.block.decorator",
            ],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
        ],
        "depth": 0,
        "data": Immutable.Map {},
      },
    },
    "selectionBefore": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
    "selectionAfter": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
  },
  "selection": Immutable.Record {
    "anchorKey": "0",
    "anchorOffset": 8,
    "focusKey": "0",
    "focusOffset": 8,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`coreBlockPlugin decorator sequence styles unstyles decorator characters when block is destroyed 1`] = `
Object {
  "content": Immutable.Record {
    "entityMap": Object {},
    "blockMap": Immutable.OrderedMap {
      "0": Immutable.Record {
        "key": "0",
        "type": "unstyled",
        "text": "##",
        "characterList": Immutable.List [
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
        ],
        "depth": 0,
        "data": Immutable.Map {},
      },
    },
    "selectionBefore": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
    "selectionAfter": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
  },
  "selection": Immutable.Record {
    "anchorKey": "0",
    "anchorOffset": 2,
    "focusKey": "0",
    "focusOffset": 2,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`coreBlockPlugin general editing Maintains block when backspacing the beginning of the block into an empty preceding block 1`] = `
Object {
  "content": Immutable.Record {
    "entityMap": Object {},
    "blockMap": Immutable.OrderedMap {
      "0": Immutable.Record {
        "key": "0",
        "type": "header-one",
        "text": "# X",
        "characterList": Immutable.List [
          Immutable.Record {
            "style": Immutable.OrderedSet [
              "core.block.decorator",
            ],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
        ],
        "depth": 0,
        "data": Immutable.Map {},
      },
    },
    "selectionBefore": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
    "selectionAfter": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
  },
  "selection": Immutable.Record {
    "anchorKey": "0",
    "anchorOffset": 0,
    "focusKey": "0",
    "focusOffset": 0,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`coreBlockPlugin general editing Maintains/expands block when forward-deleting the preceding block 1`] = `
Object {
  "content": Immutable.Record {
    "entityMap": Object {},
    "blockMap": Immutable.OrderedMap {
      "0": Immutable.Record {
        "key": "0",
        "type": "header-one",
        "text": "# Hello",
        "characterList": Immutable.List [
          Immutable.Record {
            "style": Immutable.OrderedSet [
              "core.block.decorator",
            ],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
        ],
        "depth": 0,
        "data": Immutable.Map {},
      },
    },
    "selectionBefore": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
    "selectionAfter": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
  },
  "selection": Immutable.Record {
    "anchorKey": "0",
    "anchorOffset": 2,
    "focusKey": "0",
    "focusOffset": 2,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`coreBlockPlugin headings creates a heading after typing some hashes and a space 1`] = `
Object {
  "content": Immutable.Record {
    "entityMap": Object {},
    "blockMap": Immutable.OrderedMap {
      "0": Immutable.Record {
        "key": "0",
        "type": "header-one",
        "text": "# ",
        "characterList": Immutable.List [
          Immutable.Record {
            "style": Immutable.OrderedSet [
              "core.block.decorator",
            ],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
        ],
        "depth": 0,
        "data": Immutable.Map {},
      },
    },
    "selectionBefore": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
    "selectionAfter": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
  },
  "selection": Immutable.Record {
    "anchorKey": "0",
    "anchorOffset": 2,
    "focusKey": "0",
    "focusOffset": 2,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`coreBlockPlugin unordered lists places the selection in the right place when creating a list item 1`] = `
Object {
  "content": Immutable.Record {
    "entityMap": Object {},
    "blockMap": Immutable.OrderedMap {
      "0": Immutable.Record {
        "key": "0",
        "type": "unstyled",
        "text": "",
        "characterList": Immutable.List [],
        "depth": 0,
        "data": Immutable.Map {},
      },
      "2": Immutable.Record {
        "key": "2",
        "type": "unordered-list-item",
        "text": "",
        "characterList": Immutable.List [],
        "depth": 0,
        "data": Immutable.Map {},
      },
    },
    "selectionBefore": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
    "selectionAfter": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
  },
  "selection": Immutable.Record {
    "anchorKey": "2",
    "anchorOffset": 0,
    "focusKey": "2",
    "focusOffset": 0,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`coreBlockPlugin unordered lists pressing enter after a filled list item adds another list item 1`] = `
Object {
  "content": Immutable.Record {
    "entityMap": Object {},
    "blockMap": Immutable.OrderedMap {
      "0": Immutable.Record {
        "key": "0",
        "type": "unordered-list-item",
        "text": "one",
        "characterList": Immutable.List [
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
          Immutable.Record {
            "style": Immutable.OrderedSet [],
            "entity": null,
          },
        ],
        "depth": 0,
        "data": Immutable.Map {},
      },
      "2": Immutable.Record {
        "key": "2",
        "type": "unordered-list-item",
        "text": "",
        "characterList": Immutable.List [],
        "depth": 0,
        "data": Immutable.Map {},
      },
    },
    "selectionBefore": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
    "selectionAfter": Immutable.Record {
      "anchorKey": "0",
      "anchorOffset": 0,
      "focusKey": "0",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": false,
    },
  },
  "selection": Immutable.Record {
    "anchorKey": "2",
    "anchorOffset": 0,
    "focusKey": "2",
    "focusOffset": 0,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;
